### PR TITLE
Add named_pattern and make it usable in Text_Replace

### PR DESCRIPTION
**Pull request by AdRiley**
### Pull Request Description

Creates a new type Named_Expression and allows it to be used in text_replace

image

image

image

image

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.

### CODE REVIEW

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex/Named_Pattern.enso [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+   Trailing_Whitespace
+   All_Whitespace
+   Leading_Numbers
+   Trailing_Numbers
+   Non_ASCII
+   Tabs
+   Letters
+   Numbers
+   Punctuation
+   Symbols
```
Comment (jdunkerley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex/Named_Pattern.enso]: Please add the doc comments (I already had added them back here).

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Text_Cleanse.enso [hunk 2]
```diff
@@ -1,57 +1,14 @@
...
-        Text_Cleanse.Letters -> input.replace "[a-zA-Z]" ""
-        Text_Cleanse.Numbers -> input.replace "\d" ""
-        Text_Cleanse.Punctuation -> input.replace '[,.!?():;\'\"]' ""
-        Text_Cleanse.Symbols -> input.replace "[^a-zA-Z\d\s]" ""
+   ## PRIVATE
+   Value named_pattern:Named_Pattern
+   
+   ## Apply the cleanse operation to the text.
+   apply self input:Cleansable_Text -> Any =
+      input.replace self.named_pattern.regex_pattern ""
```
Comment (jdunkerley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Text_Cleanse.enso:11]: Don't see the point of this class now.

Why not merge the apply method into `Named_Pattern` and have the convertible type in there?

File: distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Widget_Helpers.enso [hunk 3]
```diff
@@ -93,15 +93,28 @@ make_column_name_multi_selector table display:Display=Display.Always add_regex:B
...
+            builder.append (Option "<Trailing_Whitespace>" "Named_Pattern.Trailing_Whitespace")
+            builder.append (Option "<All_Whitespace>" "Named_Pattern.All_Whitespace")
+            builder.append (Option "<Leading_Numbers>" "Named_Pattern.Leading_Numbers")
+            builder.append (Option "<Trailing_Numbers>" "Named_Pattern.Trailing_Numbers")
+            builder.append (Option "<Non_ASCII>" "Named_Pattern.Non_ASCII")
+            builder.append (Option "<Tabs>" "Named_Pattern.Tabs")
+            builder.append (Option "<Letters>" "Named_Pattern.Letters")
+            builder.append (Option "<Numbers>" "Named_Pattern.Numbers")
+            builder.append (Option "<Punctuation>" "Named_Pattern.Punctuation")
+            builder.append (Option "<Symbols>" "Named_Pattern.Symbols")
```
Comment (jdunkerley) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Widget_Helpers.enso:114]: Feels odd that we angle bracket on this one and not in the Text ones but guess ok.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Column.enso [hunk 4]
```diff
@@ -1438,7 +1438,7 @@ type Column
              column.text_replace '"(.*?)"'.to_regex '($1)'
     @term make_regex_text_widget
     text_replace : Text | Regex | Column -> Text | Column -> Case_Sensitivity -> Boolean -> Column
-    text_replace self term="" new_text="" case_sensitivity=Case_Sensitivity.Sensitive only_first=False =
+    text_replace self (term : Text | Regex | Column = "") new_text="" case_sensitivity=Case_Sensitivity.Sensitive only_first=False =
```
Comment (jdunkerley) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Column.enso:1441]: Shouldn't there be similar changes to `DB_Column`?

**Review by jdunkerley — APPROVED** [review]
Couple of things to address but approving as no need to come back round once fixed.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex/Named_Pattern.enso (hunk 1 above)
Comment (AdRiley) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex/Named_Pattern.enso]: I made the type private as I don't think it should show up in the CB outside of where we use it.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Column.enso (hunk 4 above)
Comment (AdRiley) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Column.enso:1441]: I believe DB_Column had it right, but I will double check

File: distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Widget_Helpers.enso (hunk 3 above)
Comment (AdRiley) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Widget_Helpers.enso:114]: Yeah... I struggled with angle brackets or not. I'm not overly happy with it. If someone has a better idea I'm happy to change it. Angle brackets everywhere?

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Text_Cleanse.enso (hunk 2 above)
Comment (AdRiley) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Text_Cleanse.enso:11]: It defines the API of what it means to be cleansable. I almost merged it, but decided I liked it as it's own thing. Not a strong feeling though.

**Review by GregoryTravis — APPROVED** [review]

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex/Named_Pattern.enso (hunk 1 above)
Comment (AdRiley) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex/Named_Pattern.enso]: Done

**Pull request merged**